synapse/handlers/space_summary.py

```diff
@@ -378,8 +378,46 @@ async def _get_room_hierarchy(
 
                 processed_rooms.add(room_id)
             else:
-                # TODO Federation.
-                pass
+                # If a previous call got information for this room *and* it is
+                # not a space (or the max-depth has been achieved), include it.
+                if queue_entry.remote_room and (
+                    queue_entry.remote_room.get("room_type") != RoomTypes.SPACE
+                    or (max_depth is not None and current_depth >= max_depth)
+                ):
+                    room_entry = _RoomEntry(
+                        queue_entry.room_id, queue_entry.remote_room
+                    )
+                    children_room_entries: Dict[str, JsonDict] = {}
+                    inaccessible_children: Set[str] = set()
+                else:
+                    (
+                        room_entry,
+                        children_room_entries,
+                        inaccessible_children,
+                    ) = await self._summarize_remote_room_hiearchy(
+                        queue_entry,
+                        suggested_only,
+                    )
+                processed_rooms.add(room_id)
+
+                if room_entry and await self._is_remote_room_accessible(
+                    requester, queue_entry.room_id, room_entry.room
+                ):
+                    rooms_result.append(room_entry.as_json())
+
+                    # If this room is not at the max-depth, we might want to include
+                    # children.
+                    if max_depth is None or current_depth < max_depth:
+                        room_queue.extendleft(
```

Comment:
Because Python! From [their docs](https://docs.python.org/3/library/collections.html#collections.deque):

> Deques are a generalization of stacks and queues (the name is pronounced “deck” and is short for “double-ended queue”). Deques support thread-safe, memory efficient appends and pops from either side of the deque with approximately the same O(1) performance in either direction.

Reply:
sure, but you could also use a `list`, which has O(1) append and pop at one end.

Reply:
Yeah, switching to a list is a bit easier to understand. Good call! 👍